### ISSUE TITLE
Parallel Scaling of Libint2 One-Electron Integrals

### ISSUE THREAD

**Issue by zachglick** [psi4/psi4#2491]
The one-electron integrals (OEIs) in psi4 are in the process of being overhauled. The (old) native psi4 Obara-Saika code was recently replaced with an interface to the (new) libint2 library (#2388). I've found that in my specific use case, while the new OEI code is serially faster than the old OEI code, the new OEI code scales surprisingly poorly with multiple cores. 

### Background

I'm in the process of implementing semi-numerical exchange for psi4. The cost of this algorithm is dominated by the calculation of OEIs with the following functional form:
```
(u | (r - r_g)^{-1} | v),
```
where `u` and `v` are basis functions, and `r_g` is a point in space on some quadrature. A naive semi-numerical exchange code computes these integrals over all `u`, `v`, and `r_g`, requiring O(N^3) work.

In current psi4, integrals of this type can be computed with either the old OS code (via the `MultipolePotentialInt` class) or the new libint2 interface (via the `ElectrostaticInt` class). I've attached a plugin that computes a large number of these OEI integrals with both classes, modeled after a naive semi-numerical exchange code. Here's the pseudocode of this plugin:

```
# pragma omp parallel for
for grid_point_block in quadrature (B)
  
  thread = omp_get_thread()
  
  for bra in basis (u)
    for ket in basis (v)
      for grid point in B (g)
      
        // integral_computer[thread] is either an ElectrostaticInt or MultipolePotentialInt
        integral_computer[thread].compute(u, v, g)
```

Note that the above pseudocode is parallelized over blocks of ~64 grid points. This pseudocode follows the description of the original chain of spheres (COSX) semi-numerical exchange algorithm described [here](https://www.sciencedirect.com/science/article/pii/S0301010408005089).

### Results

Below are timings of the naive semi-numerical integral loop using both the old psi4 integrals and the new libint2 integrals. This benchmark was performed on a butane molecule in a cc-pVDZ basis set and a standard 302/75 quadrature grid. Wall times are the minimum of five trials.

| Cores | Old Integral<br/>Wall Time (s) | Old Integral<br/>Parallel Speedup |  New Integral<br/>Wall Time (s) | New Integral<br/>Parallel Speedup |
| :---     |  :---     | :---      |  :---    | :---    |
| 1         | 191.7    | ---       | 105.3  | ---     |
| 2         | 96.1     | x2.00  | 72.6    | x1.45 |
| 4         | 48.4    | x3.96  | 71.5     | x1.47 |
| 8         | 24.3     | x7.88  | 74.5     | x1.41 |
| 18       | 11.5      | x16.62 | 72.8     | x1.45 |

Note that with one core, the new libint2 integrals (105.3s) are faster than the old OS integrals (191.7s) by almost a factor of two, which is to be expected. Part of the reason for deprecating the old, hand-coded integral code for the libint2 interface is that libint2 is a specialized library and therefore should have more efficient routines.

The integral routine using the old OS code scales almost perfectly all the way through 18 cores. This is also to be expected: the work being done in the integral routine is embarrassingly parallel, and threads don't share any information or intermediates.

Surprisingly, when the same integrals are computed with libint2 instead of the old OS code, the parallel scaling suffers drastically. We see no benefit gained from using more than two cores, and even a calculation with two cores only achieves a 1.45x speedup. When I use all 18 cores on my workstation, the new libint2 integrals are nearly 7(!) times slower than the old OS code.

What makes this particularly confusing is that none of the usual culprits associated with poor parallel scaling (bad scheduling/work balance, high overhead, etc.) apply here. This is because we see perfect parallel scaling with the old OS code using all of the same conditions. The only difference between the two computations is that I've switched the source of the integrals.

The only explanation for this behavior I can come up with is that the libint2 library might have some kind of global lock that prevents independent threads (with their own independent `ElectrostaticInt` objects) from computing integrals in parallel. I don't know enough about how libint2 works to justify this idea. I'm very curious to hear other thoughts on this scaling issue.

### Plugin

The plugin (attached) requires adding `PSI_API` to a few classes (DFTGrid, MultipolePotentialInt). The following function is also necessary in `psi4/libmints/electrostatic.cc`:
```
void ElectrostaticInt::set_origin(const Vector3& _origin) {
    origin_ = _origin;
    std::vector<std::pair<double, std::array<double, 3>>> pcs;
    pcs.push_back({-1.0, {origin_[0], origin_[1], origin_[2]}});
    engine0_->set_params(pcs);
}
```

[plugin.cc.zip](https://github.com/psi4/psi4/files/8326964/plugin.cc.zip)

**Comment by TiborGY** [comment]
I have no idea if this is the cause here, but there is a somewhat obscure effect called "false sharing" where CPU cache interference between threads can wreck parallel scaling.

**Comment by hokru** [comment]
That Libint2 OEI got introduced in https://github.com/psi4/psi4/pull/2388

**Comment by jturney** [comment]
Are you able to run this through VTune?

**Comment by hokru** [comment]
We are still using an old libint2 until PR #2413 gets in.

In the libint2 changelog i see:
 `fixed excessive lock contention when computing 1-e Coulomb ("nuclear") ints on many threads`
https://github.com/evaleev/libint/blob/master/CHANGES#L16
https://github.com/evaleev/libint/commit/916052a656081dfe05ae902d48aae3c513d99495

Maybe this is related?

**Comment by zachglick** [comment]
@TiborGY That's a good idea! False sharing is a reasonable thing to think about. From what I've just now read, it sounds like false sharing occurs when two threads repeatedly write to nearby array elements, invalidating cached values of those elements. In this minimal test case, the threads aren't writing to any shared data, so I don't think false sharing applies here. However, it's something I'll look out for in the actual semi-numerical exchange implementation.

@jturney I haven't gotten desperate enough to learn VTune yet, but I'm getting close.

@hokru Thank you! The changelog entry would perfectly explain this strange behavior. I'll wait for #2413 before doing any more investigation.

**Comment by zachglick** [comment]
Update: This performance issue was fixed by #2413, which upgraded libint2 from 2.6.0 to 2.7.1. For completeness, here are the timings on the above test case:

| Cores | Psi4 OS<br/>Wall Time (s) | Psi4 OS<br/>Parallel Speedup |  Libint 2.6.0<br/>Wall Time (s) | Libint 2.6.0<br/>Parallel Speedup |  Libint 2.7.1<br/>Wall Time (s) | Libint 2.7.1<br/>Parallel Speedup |
| :---     |  :---     | :---      |  :---    | :---    |  :---    | :---    |
| 1         | 191.7    | ---       | 105.3  | ---     | 99.4  | ---     |
| 2         | 96.1     | x2.00  | 72.6    | x1.45 | 49.9    | x1.99 |
| 4         | 48.4    | x3.96  | 71.5     | x1.47 | 25.2     | x3.94 |
| 8         | 24.3     | x7.88  | 74.5     | x1.41 | 13.0     | x7.66 |
| 18       | 11.5      | x16.62 | 72.8     | x1.45 | 6.4     | x15.45 |
